### ISSUE TITLE
Ninja parses empty String to NULL problem

### ISSUE THREAD

**Issue by rrarrarra** [ninjaframework/ninja#588]
Problem:
Ninja does not correctly parse the object ("MyObject") given in the controller-method.
An empty String will be parsed to NULL instead of empty String ("").
But this should not be NULL! It should be the empty String ("").

Given:
```java
public class MyObject {
   public String bla;
}
```

and
```java
public class ApplicationController {
    public Result index( Context context, MyObject myObject) {
	System.out.println("test1: " + context.getParameter("bla"));
	System.out.println("test2: " + myObject.bla);
    }
}
```

Sending form-POST to Controller-Method 
```html
<form method="POST">
   <input type="text" name="bla" value="">
</form>
```

will result in:

```console
test1: 
test2: null
```

Thanks for help/fix.

**Comment by raphaelbauer** [comment]
Hey @rranke, thanks for your report!

Actually that is a feature, not a bug. Full discussion and the PR can be found here: https://github.com/ninjaframework/ninja/pull/573

Does that make sense to you, or is it completely outrageous?

**Comment by rrarrarra** [comment]
Thanks for the link to #573 !

My opinion about this:

1.)
I think, for to be able to handle each situation on server-side, the best solution would be:


`If parameter IS-NOT in the Request`
 ->
 `null should be the parsing-result`

`If parameter IS in the Request AND has no value`
-> 
`The empty String ("") should be the parsing-result`

`If parameter IS in the Request AND has a value` 
-> 
`The value-as-a-String should be the parsing-result`


2.) 
I cannot use the benefit of ninjas-object-parsing-feature with the current solution.
I have no chance to get the intention of the clients POST:

- Does the client wanted to initialize a field and set it to the "empty string"?
or
- Does the client not want to change the value of the field at all?
or 
- Has the client just forgotten to set the value of a field?
or
- Is there a client-side bug, maybe just forgotten to add the field to the form?

For example, my Database uses some logic based on "null" and "empty String" text-fields.
I have a Database text-field, that is set to NULL, to have a marker, that is not yet initialized.
If it is not NULL any more, then I know, that the field was initialized.

With the current solution, when the client wants to set a value to the "empty String", I cannot determine this particular intention.
On server side, it looks always like, that the client has just forgotten to send the parameter.
-> I cannot easily initialize a field to the "empty String" value.

Or am I missing something here?



3.) 
Either way, it should be consistent all over the framework. So, they both should print out the same value in every case:
```java
	System.out.println("test1: " + context.getParameter("bla"));
	System.out.println("test2: " + myObject.bla);
```

Or What is the benefit of having two different outputs here, in the current implementation of the Framework?

**Comment by raphaelbauer** [comment]
Thanks for the clarification! Pinging @jjlauer and @jlannoy - what's your opinion?

**Comment by jlannoy** [comment]
> For example, my Database uses some logic based on "null" and "empty String" text-fields.

But how can you make the difference between null string or an empty string? The default behavior for HTML forms is that as soon as your html form contains the field, it will sends an (empty) parameter. Are you dynamically removing or adding fields?

When Ninja tries to parse objects, it's a little bit more clever than simply checking for parameters existence. All type conversions are made, and then it's considered that if the user did not enters anything in a field, the string is null. In fact, as for a long, an integer, or any other type... This is why a "getParameter" is different. If you really want to see that your user entered "aaa" for a long, you get the parameter, not the long attribute of your DTO. That's the logic besides the object parser.

Btw, you can easily set it up the way you want, just by creating your own String parser and overriding the default one :
```java
    public class StringParamParser implements ParamParser<String> {    
        @Override
        public String parseParameter(String field, String parameterValue, Validation validation) {
            if (parameterValue == null /* || parameterValue.isEmpty() */ || validation.hasViolation(field)) {
                return null;
            } else {
                return parameterValue;
            }
        }

        @Override
        public Class<String> getParsedType() {
            return String.class;
        }
    }
```

And adding it inside your conf.Module.configure() :

```java
Multibinder parsersBinder = Multibinder.newSetBinder(binder(), ParamParser.class);
parsersBinder.addBinding().to(StringParamParser.class);
```

**Comment by rrarrarra** [comment]
Example:

- Normal value:
```html
<form>
 <input type="text" name="locality" value="Berlin">
 <input type="text" name="street" value="Schneider Str. 1">
</form>
```
```java
String locality = "Berlin";
String street = "Schneider Str. 1"; //<--- OK
```

- Empty String value:
```html
<form>
 <input type="text" name="locality" value="Berlin">
 <input type="text" name="street" value="">
</form>
```
```java
String locality = "Berlin";
String street = null;  //<------ PROBLEM! current implementation sets it to null :(
String street = "";  //<------ i want to have that! How to get it?
```

- NULL String value:
```html
<form>
 <input type="text" name="locality" value="Berlin">
 <!-- <input type="text" name="street" value=""> -->
</form>
```
```java
String locality = "Berlin";
String street = null; //<--- OK!
```

Thanks @jlannoy , for clarifications and for providing a solution!!

Has anybody checked out how other frameworks  (e.g. play) behave?

This whole thing is not a deal-breaker for me. I can live with that ;)
But I just wanted to point out, that with the current implementation, it is pretty hard (or impossible?) to  get empty-Strings automatically parsed into a java-object. 

Or do you @jlannoy know a way how to send a parameter that gets automatically parsed into javas:
```java
String street = "";  //<------ i want to have that! How to get it?
```

**Comment by jlannoy** [comment]
The fact that you comment a field to get a null value server side seems me odd. It's doesn't represent any human behavior, but a technical one. I don't say Ninja proposes the best solution, we tried to discuss about it and now you're the first one that point it the other way :)

But, again... It's not pretty hard or impossible ;)
With the 17 lines I gave you, **all** your Strings for **all** your objects will be parsed as you want it to. You'll not have to modify anything in the manner the parameters are sent. Check the 3rd line of the StringParamParser, it defines if empty strings must be considered as null. As I commented it for you, you'll have what you want.
